mbedtls/tests/async_session.rs

```diff
@@ -0,0 +1,362 @@
+/* Copyright (c) Fortanix, Inc.
+ *
+ * Licensed under the GNU General Public License, version 2 <LICENSE-GPL or
+ * https://www.gnu.org/licenses/gpl-2.0.html> or the Apache License, Version
+ * 2.0 <LICENSE-APACHE or http://www.apache.org/licenses/LICENSE-2.0>, at your
+ * option. This file may not be copied, modified, or distributed except
+ * according to those terms. */
+
+#![cfg(not(target_env = "sgx"))]
+extern crate mbedtls;
+
+use std::future::Future;
+use std::pin::Pin;
+use std::sync::Arc;
+
+use mbedtls::pk::Pk;
+use mbedtls::rng::CtrDrbg;
+use mbedtls::ssl::async_io::AsyncIoAdapter;
+use mbedtls::ssl::config::{Endpoint, Preset, Transport};
+use mbedtls::ssl::io::IoCallback;
+use mbedtls::ssl::{Config, Context, Version};
+use mbedtls::x509::{Certificate, VerifyError};
+use mbedtls::Error;
+use mbedtls::Result as TlsResult;
+
+use tokio::io::{AsyncRead, AsyncReadExt, AsyncWrite, AsyncWriteExt};
+
+mod support;
+use support::entropy::entropy_new;
+use support::keys;
+
+// TODO: Add unified interface for TCP and UDP like `TransportType` in `client_server.rs`
+
+async fn client<C: AsyncRead + AsyncWrite + Unpin + 'static, T>(
```

Comment:
If you're only supporting TcpStream I don't see any value in having this generic.

Reply:
removed